File: src/com/designpatterns/builder/PersonBuilder.java
package com.designpatterns.builder;

import java.util.Date;

public class PersonBuilder<T extends Person<T>> {
    protected Person person;
    PersonBuilder(){
        person=new Person();
    }
    public PersonAddressBuilder address(){
        return new PersonAddressBuilder(person);
    }
    public PersonWorkBuilder work(){
        return new PersonWorkBuilder(person);
    }
    public PersonBuilder<T> withName(String name){
        person.name=name;
        return this;
    }
    public PersonBuilder<T> withDob(Date date){
        person.dob=date;
        return this;
    }
    public PersonBuilder<T> withAge(int age){
        person.age=age;
        return this;
    }
    public Person build(){
        return person;
    }

}

File: js/Builder/PersonBuilder.js
const Person = require("./Person");

class PersonBuilder {
  constructor() {}
  address() {
    return new PersonAddressBuilder(this);
  }
  work() {
    return new PersonWorkBuilder(this);
  }
  withName(name) {
    this.name = name;
    return this;
  }
  withAge(age) {
    this.age = age;
    return this;
  }
  withDob(date) {
    this.dob = date;
    return this;
  }
  build() {
    return new Person(this);
  }
}

class PersonAddressBuilder extends PersonBuilder {
  constructor(person) {
    super(person);
    console.log(person);
    this._person=person;
  }
  withCity(city) {
    this._person.city = city;
    return this;
  }
  withState(state) {
    this._person.state = state;
    return this;
  }
  withPinCode(pinCode) {
    this._person.pincode = pinCode;
    return this;
  }
  build(){
      return new Person(this._person);
  }
}

class PersonWorkBuilder extends PersonBuilder {
  constructor(person) {
    super();
    this._person = person;
  }
  withCompany(name) {
    this._person.company = name;
    return this;
  }
  withOfficeCity(city) {
    this._person.officeCity = city;
    return this;
  }
  withDesignation(designation) {
    this._person.designation = designation;
    return this;
  }
  build(){
      return new Person(this._person);
  }
}

module.exports = {
  PersonWorkBuilder: PersonWorkBuilder,
  PersonBuilder: PersonBuilder,
  PersonAddressBuilder: PersonAddressBuilder,
};
